docs/book/command-params.md

```diff
@@ -0,0 +1,203 @@
+to simply write interactive commands.
+
+## Using input parameters
+
+Internally input parameters behaves like options with the following modifications:
+
+- all parameters are optional by default,
+- in interactive mode, when getting value parameter value, and the value
+  is not provided with the command call, user will be asked to provide the value,
+- in non-interactive mode, default value is used instead, if the parameter value
+  is not supplied, exception will be thrown,
+- values provided in command call are validated the same way as question asked
+  to provide the value by user,
+- if value of the parameter is provided by user is set back in input options,
+  so it can be passed to the next [command in chain](commands-chain.md) (if defined),
+- there is no shortcut version of parameter, as we have for option.
+
+Please consider the following command:
+
+```php
+use Laminas\Cli\Input\InputParam;
+use Laminas\Cli\Input\InputParamTrait;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+final class HelloCommand extends Command
+{
+    use InputParamTrait;
+
+    public static $defaultName = 'example:hello';
+
+    protected function configure() : void
+    {
+        $this->setName(self::$defaultName);
+        $this->addParam('name', 'Your name', InputParam::TYPE_STRING);
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output) : int
+    {
+        $name = $this->getParam('name');
+        $output->writeln('Hello ' . $name . '!');
+
+        return 0;
+    }
+}
+```
+
+As you see it is very similar to adding option. As noted before there is no shortcut,
+so we always must use the full name.
+
+Now, let what happens when we call the command:
+
+```console
+$ vendor/bin/laminas example:hello
+Your name:
+ > Michal
+
+Hello Michal!
+```
+
+or, if you provide the param value with the command call:
+
+```console
+$ vendor/bin/laminas example:hello --name=Michal
+
+Hello Michal!
+```
+
+As you can see, there is no question asked when the parameter value is provided on command call.
+
+## Build-in types
+
+### InputParam::TYPE_BOOL
+
+```php
+$this->addParam(
+    'enable',
+    'Enable something',
+    InputParam::TYPE_BOOL,
+    false, // if the param is required, ignored for TYPE_BOOL
+    false // default value in interactive mode: true or false
+);
+```
+
+If the option is not set in command line call, in interactive mode use is going to be asked for confirmation:
+
+```console
+Enable something? [y/N]
+ > ...
+```
+
+Value of this parameter is normalized to `boolean` type.
+
+### InputParam::TYPE_INT
+
+```php
+$this->addParam(
+    'age',
+    'Your age',
+    InputParam::TYPE_INT,
+    true, // if the param is required (answer in interactive mode, if param value is not supplied)
+    null, // no default value
+    [
+        'min' => 0,
+        'max' => 120,
+    ]
```

Comment:
I think my biggest issue here is just the sheer number of arguments. As I've noted in another comment, _this is a problem in the symfony/console API in general_; I just wish we could do something that might offer an alternative API to simplify it.

Now that I'm thinking about it, I wonder if we might also allow this:

```php
$param = new InputParam('module', InputParam::TYPE_STRING);
$param->setDescription(/* ... */);
$param->setShortcut('m');
$param->addValidator('pattern', '/^[A-Z][a-zA-Z0-9_]*$/');
$param->setRequired(false);
$this->addParam($param);
```

Essentially, populate it first, and then add it. This gives type-safety, and an easier way to provide the various optional values, and you could then pull data from the param to populate the `InputOption`. By doing this, we also no longer have to understand what values are possible for each argument; instead, we have clearly defined methods with clearly defined signatures.

If each of those methods returns `self`, it could even look like this:

```php
$this->addParam((new InputParam('module', InputParam::TYPE_STRING))
    ->setDescription('The module to create')
    ->setShortcut('m')
    ->addValidator('pattern', '/^[A-Z][a-zA-Z0-9_]*$/')
    ->setRequired(false)
);
```

Which, while more verbose, makes what each line is adding to the parameter far easier to understand for somebody looking at the code who doesn't know what `addParam()` expects for arguments.